drivers/adc/adc_b91.c

```diff
@@ -0,0 +1,464 @@
+		data->resolution_divider = 1;
+		break;
+	case 12:
+		adc_set_resolution(ADC_RES12);
+		data->resolution_divider = 4;
+		break;
+	case 10:
+		adc_set_resolution(ADC_RES10);
+		data->resolution_divider = 16;
+		break;
+	case 8:
+		adc_set_resolution(ADC_RES8);
+		data->resolution_divider = 64;
+		break;
+	default:
+		LOG_ERR("Selected ADC resolution is not supported.");
+		return -EINVAL;
+	}
+
+	/* Save buffer */
+	data->buffer = sequence->buffer;
+
+	/* Start ADC conversion */
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+/* Main ADC Acquisition thread */
+static void adc_b91_acquisition_thread(const struct device *dev)
+{
+	int16_t adc_code;
+	struct b91_adc_data *data = dev->data;
+
+	while (true) {
+		/* Wait for Acquisition semaphore */
+		k_sem_take(&data->acq_sem, K_FOREVER);
+
+		/* Wait for ADC data ready */
+		while ((analog_read_reg8(AREG_ADC_DATA_STATUS) & ADC_DATA_READY)
+				!= ADC_DATA_READY) {
+		}
+
+		/* Perform read */
+		adc_code = (adc_b91_get_code() / data->resolution_divider);
+		if (!data->differential) {
+			/* Sign bit is not used in case of single-ended configuration */
+			adc_code = adc_code * 2;
+
+			/* Do not return negative value for single-ended configuration */
+			if (adc_code < 0) {
+				adc_code = 0;
+			}
+		}
+		*data->buffer++ = adc_code;
+
+		/* Power off ADC */
+		adc_power_off();
+
+		/* Release ADC context */
+		adc_context_on_sampling_done(&data->ctx, dev);
+	}
+}
+
+/* ADC Driver initialization */
+static int adc_b91_init(const struct device *dev)
+{
+	struct b91_adc_data *data = dev->data;
+
+	k_sem_init(&data->acq_sem, 0, 1);
+
+	k_thread_create(&data->thread, data->stack,
+			CONFIG_ADC_B91_ACQUISITION_THREAD_STACK_SIZE,
+			(k_thread_entry_t)adc_b91_acquisition_thread,
+			(void *)dev, NULL, NULL,
+			CONFIG_ADC_B91_ACQUISITION_THREAD_PRIO,
+			0, K_NO_WAIT);
+
+	adc_context_unlock_unconditionally(&data->ctx);
+
+	return 0;
+}
+
+/* API implementation: channel_setup */
+static int adc_b91_channel_setup(const struct device *dev,
+				 const struct adc_channel_cfg *channel_cfg)
+{
+	adc_ref_vol_e vref_internal_mv;
+	adc_sample_freq_e sample_freq;
+	adc_pre_scale_e pre_scale;
+	adc_sample_cycle_e sample_cycl;
+	adc_input_pin_def_e input_positive;
+	adc_input_pin_def_e input_negative;
+	struct b91_adc_data *data = dev->data;
+	const struct b91_adc_cfg *config = dev->config;
+
+	/* Check channel ID */
+	if (channel_cfg->channel_id > 0) {
+		LOG_ERR("Only channel 0 is supported.");
+		return -EINVAL;
+	}
+
+	/* Check reference */
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Selected ADC reference is not supported.");
+		return -EINVAL;
+	}
+
+	/* Check internal reference */
+	switch (config->vref_internal_mv) {
+	case 900:
+		vref_internal_mv = ADC_VREF_0P9V;
+		break;
+	case 1200:
+		vref_internal_mv = ADC_VREF_1P2V;
+		break;
+	default:
+		LOG_ERR("Selected reference voltage is not supported.");
```

Comment:
Wouldn't a build assertion be more suitable for signaling this misconfiguration?

Reply:
This case should never happen. We can not use build assertion inside the switch, I prefer to be consistent with the rest configs here and with other ADC drivers.
No changes here.